### ISSUE TITLE
Enhancement request: simulate command execution in test harness

### ISSUE THREAD

**Issue by weiiwang01** [canonical/operator#932]
The testing module's test harness provides various simulations to facilitate charm execution during unit testing. However, the command execution performed by `ops.model.Container.exec` is currently not simulated. Consequently, when the charm code invokes `ops.model.Container.exec` during unit tests, it attempts to execute commands on the development machine. In many cases, the development environment lacks the required commands and necessary execution environments. This makes code patching required to mock command execution during unit testing.

To address this limitation, it would great for the test harness to offer an API that allows registering callback functions for simulating command execution. After the harness has started, command executions can be dispatched to pre-registered callback functions.

Here's an example of a crude implementation in our unit tests for reference:

https://github.com/canonical/flask-k8s-operator/blob/979f3aaf0b8f120a874d1ad9f039983b7ffc37eb/tests/unit/conftest.py#L20-L68
```python
def inject_register_command_handler(monkeypatch: pytest.MonkeyPatch, harness: Harness):
    """A helper function for injecting an implementation of the register_command_handler method."""
    handler_table: dict[str, typing.Callable[[list[str]], tuple[int, str, str]]] = {}

    class ExecProcessStub:
        """A mock object that simulates the execution of a process in the container."""

        def __init__(self, command: list[str], exit_code: int, stdout: str, stderr: str):
            """Initialize the ExecProcessStub object."""
            self._command = command
            self._exit_code = exit_code
            self._stdout = stdout
            self._stderr = stderr

        def wait_output(self):
            """Simulate the wait_output method of the container object."""
            if self._exit_code == 0:
                return self._stdout, self._stderr
            raise ExecError(
                command=self._command,
                exit_code=self._exit_code,
                stdout=self._stdout,
                stderr=self._stderr,
            )

    def exec_stub(command: list[str], **_kwargs):
        """A mock implementation of the `exec` method of the container object."""
        executable = command[0]
        handler = handler_table[executable]
        exit_code, stdout, stderr = handler(command)
        return ExecProcessStub(command=command, exit_code=exit_code, stdout=stdout, stderr=stderr)

    def register_command_handler(
        container: Container | str,
        executable: str,
        handler=typing.Callable[[list[str]], typing.Tuple[int, str, str]],
    ):
        """Registers a handler for a specific executable command."""
        container = (
            harness.model.unit.get_container(container)
            if isinstance(container, str)
            else container
        )
        handler_table[executable] = handler
        monkeypatch.setattr(container, "exec", exec_stub)

    monkeypatch.setattr(
        harness, "register_command_handler", register_command_handler, raising=False
    )
```

https://github.com/canonical/flask-k8s-operator/blob/979f3aaf0b8f120a874d1ad9f039983b7ffc37eb/tests/unit/conftest.py#L79-L102
```python
    def python_cmd_handler(argv: list[str]) -> ExecResult:
        """Handle the python command execution inside the Flask container."""
        match argv:
            case [
                "python3",
                "-m",
                "gunicorn",
                "-c",
                "/srv/flask/gunicorn.conf.py",
                "app:app",
                "--check-config",
            ]:
                if flask_container.exists(
                    "/srv/flask/gunicorn.conf.py"
                ) and not flask_container.isdir("/srv/flask/gunicorn.conf.py"):
                    return ExecResult(0, "", "")
                return ExecResult(1, "", "")
            case _:
                raise RuntimeError(f"unknown command: {argv}")

    inject_register_command_handler(monkeypatch, harness)
    harness.register_command_handler(  # type: ignore # pylint: disable=no-member
        container=flask_container, executable="python3", handler=python_cmd_handler
    )
```

**Comment by benhoyt** [comment]
Yeah, I think this is an important feature missing from Harness right now. I've added the "needs spec" label as I think there are a few details to iron out here: naming, should it be a callback like you've shown or perhaps a `with` context manager, should the first arg / executable be separated out, should we support all of exec's args & features, and so on.

I'd like to be able to add this this cycle, but not sure I'll have the time (it's not on the roadmap). @weiiwang01 What would you think about putting a spec together to kick off the discussion? I'd be happy to do it over a PR too, but it might be better to have a bit more back and forth before coding.

**Comment by benhoyt** [comment]
This was spec'd in [OP039](https://docs.google.com/document/d/1V7nzFzSOIMVlPF-HCvTZZyWOGlHnCS__RyL5352P2e4/edit), implemented in #993, and released in [ops 2.6.0](https://github.com/canonical/operator/releases/tag/2.6.0). Thanks for your work on this!
